tests/test_resolver.py

```diff
@@ -0,0 +1,129 @@
+    keys_strategy = st.lists(st.sampled_from(list(input_dict.keys())), unique=True)
+
+    def build_dict(keys):
+        return {key: input_dict[key] for key in keys}
+
+    return keys_strategy.map(build_dict)
+
+
+# The deps in each category should be disjoint
+
+# non locally installed deps
+non_locally_installed_deps = ["pandas", "numpy", "other"]
+non_locally_installed_strategy = st.lists(
+    st.sampled_from(non_locally_installed_deps), unique=True
+)
+
+# locally installed deps
+locally_installed_deps = {
+    "setuptools": [
+        "_distutils_hack",
+        "pkg_resources",
+        "setuptools",
+    ],
+    "pip": ["pip"],
+    "isort": ["isort"],
+}
+locally_installed_strategy = dict_subset_strategy(locally_installed_deps)
+
+# user-defined deps
+user_defined_deps = {"apache-airflow": ["airflow"]}
+user_defined_strategy = dict_subset_strategy(user_defined_deps)
+
+# user-defined mapping
+user_defined_mapping = "\n".join(
+    [f'{dep} = ["{",".join(imports)}"]' for dep, imports in user_defined_deps.items()]
+)
+
+user_mappings = st.one_of(st.none(), st.just(user_defined_mapping))
+
+
+def generate_expected_resolved_deps(
+    locally_installed_deps: Optional[Dict[str, List[str]]] = None,
+    non_locally_installed_deps: Optional[List[str]] = None,
+    user_defined_deps: Optional[Dict[str, List[str]]] = None,
+):
+    ret = {}
+    if locally_installed_deps:
+        ret.update(
+            {
+                dep: Package(
+                    dep,
+                    {DependenciesMapping.LOCAL_ENV: set(imports)},
+                )
+                for dep, imports in locally_installed_deps.items()
+            }
+        )
+    if user_defined_deps:
+        ret.update(
+            {
+                dep: Package(
+                    dep,
+                    {DependenciesMapping.USER_DEFINED: set(imports)},
+                )
+                for dep, imports in user_defined_deps.items()
+            }
+        )
+    if non_locally_installed_deps:
+        ret.update(
+            {
+                dep: Package(dep, {DependenciesMapping.IDENTITY: {dep}})
+                for dep in non_locally_installed_deps
+            }
+        )
+    return ret
+
+
+# Suppressing the warning on function scoped fixtures: even if the tmp writing
+# fixture only runs once for all the test cases, it will be writing the same
+# content
+@given(
+    user_mapping=user_mappings,
+    non_installed_deps=non_locally_installed_strategy,
+    user_deps=user_defined_strategy,
+    installed_deps=locally_installed_strategy,
+)
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+def test_resolve_dependencies__generates_expected_mappings(
+    user_deps,
+    installed_deps,
+    non_installed_deps,
+    user_mapping,
+    write_tmp_files,
+):
+
+    # The case where there are user-defined deps but no user-defined mapping
+    # provided is not valid
+    assume(not (len(user_deps) > 0 and user_mapping is None))
```

Comment:
WIP - I will have a comment here

Reply:
Now that this is rebased on the changes that include having both a file-defined mapping and a toml-defined mapping, there are three different variables to distinguish:
- A file-defined user mapping + a strategy that either includes that mapping in the test input or passes `None`
- A toml user mapping + a strategy to either include it or not in the test input
- A list of user deps. This list is part of the deps that would be passed to the resolver. It has no overlap with the other categories of deps (locally-installed and other deps). And it includes the deps that should be resolved from the user-provided mapping.

It is not valid to have some deps that need to be resolved from the user-defined mapping while no user-defined mapping is provided. This could happen when the strategies for both the file-defined and the toml-defined user mappings produce `None`, but `user_deps` is not an empty list (it contains e.g. `apache-airflow`, which would not be resolved as a user-defined dependency in this case).